src/graph/transform/to_bipartite.h

```diff
@@ -49,6 +60,26 @@ std::tuple<HeteroGraphPtr, std::vector<IdArray>> ToBlock(
     HeteroGraphPtr graph, const std::vector<IdArray>& rhs_nodes,
     bool include_rhs_in_lhs, std::vector<IdArray>* lhs_nodes);
 
+/**
+ * @brief A warpper function shared by CPU and GPU ```ToBlock```
+ * which deal with the common preprocess and postprocess work of them.
+ *
+ * @tparam IdType The type to use as an index.
+ * @param graph The graph from which to extract the block.
+ * @param rhs_nodes The destination nodes of the block.
+ * @param include_rhs_in_lhs Whether or not to include the
+ * destination nodes of the block in the sources nodes.
+ * @param [in/out] lhs_nodes The source nodes of the block.
+ * @param MappingIdsFunc  The function to get mapped ids from original ids.
+ *
+ * @return The block and the induced edges.
+ */
+template <typename IdType>
+std::tuple<HeteroGraphPtr, std::vector<IdArray>> ToBlockProcess(
```

Comment:
It is a good point and thanks for the detailed code explanation. But as `ToBlock` is more about mapping ids, changing the name to `MapNodeIds` may be not appropriate.  
And it could be hard to dispatch `IdMapper` here because it needs to be inserted in the middle of the `ToBlock ` logic, and if we run it separately outside, there need to be a lot of intermediate variables and make the code looks more complex.

Reply:
Sorry, hopefully the below example is more clear how to avoid the double dispatch logic here. Instead of passing in the function as a template parameter, we could just modify the one place in `ToBlockProcess()` where it is invoked, and instead dispatch based on the device type template parameter already provided (in the below example I use `MapSubgraphIds()`, but please choose whatever name you feel is most appropriate):

```c++
template<typename XPU, typename IdType>
std::tuple<HeteroGraphPtr, std::vector<IdArray>> ProcessToBlock(
    HeteroGraphPtr graph, const std::vector<IdArray> &rhs_nodes,
    bool include_rhs_in_lhs, std::vector<IdArray> *const lhs_nodes_ptr) {
...
  std::vector<int64_t> num_nodes_per_type(num_ntypes * 2);
  // Populate RHS nodes from what we already know.
  for (int64_t ntype = 0; ntype < num_ntypes; ++ntype) {
    num_nodes_per_type[num_ntypes + ntype] = rhs_nodes[ntype]->shape[0];
  }
  std::vector<IdArray> new_lhs;
  std::vector<IdArray> new_rhs;

  std::tie(new_lhs, new_rhs) = MapSubgraphIds<XPU, IdType>(
          graph, include_rhs_in_lhs, num_ntypes, ctx, maxNodesPerType, edge_arrays,
          src_nodes, rhs_nodes, lhs_nodes_ptr, &num_nodes_per_type);

  std::vector<IdArray> induced_edges;
  induced_edges.reserve(num_etypes);
  for (int64_t etype = 0; etype < num_etypes; ++etype) {
    if (edge_arrays[etype].id.defined()) {
      induced_edges.push_back(edge_arrays[etype].id);
    } else {
      induced_edges.push_back(
          aten::NullArray(DGLDataType{kDGLInt, sizeof(IdType) * 8, 1}, ctx));
    }
  }
...
}
```

And then rather than specialize `ToBlock()` for the device types, we could have it directly call `ProcessToBlock()` (or just rename `ProcessToBlock()` to `ToBlock()`):
```c++
template<typename XPU, typename IdType>
std::tuple<HeteroGraphPtr, std::vector<IdArray>> ToBlock(
    HeteroGraphPtr graph, const std::vector<IdArray> &rhs_nodes,
    bool include_rhs_in_lhs, std::vector<IdArray> *const lhs_nodes) {
  return ProcessToBlock<XPU, IdType>(graph, rhs_nodes, include_rhs_in_lhs, lhs_nodes);
}
```

We would no longer need the `ToBlockCPU()`, `ToBlockGPU32()`, or `ToBlockGPU64()` functions as the specialization would happen at the `MapSubgraphIds()` function.

Given #5305 is merged to master, I'm also fine with this PR as is.